src/platform/cc13xx_26xx/FactoryDataProvider.cpp

```diff
@@ -0,0 +1,333 @@
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FactoryDataProvider::GetProductAttestationIntermediateCert(MutableByteSpan & outBuffer)
+{
+    ReturnErrorCodeIf(outBuffer.size() < mFactoryData.pai_cert.len, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(!mFactoryData.pai_cert.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(outBuffer.data(), mFactoryData.pai_cert.data, mFactoryData.pai_cert.len);
+    outBuffer.reduce_size(mFactoryData.pai_cert.len);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FactoryDataProvider::SignWithDeviceAttestationKey(const ByteSpan & messageToSign, MutableByteSpan & outSignBuffer)
+{
+    Crypto::P256ECDSASignature signature;
+    Crypto::P256Keypair keypair;
+
+    VerifyOrReturnError(IsSpanUsable(outSignBuffer), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(IsSpanUsable(messageToSign), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(outSignBuffer.size() >= signature.Capacity(), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Extract public key from DAC cert.
+    uint8_t dacBuf[600] = { 0 };
+    MutableByteSpan dacCertSpan(dacBuf);
+    memcpy(dacCertSpan.data(), mFactoryData.dac_cert.data, mFactoryData.dac_cert.len);
+    dacCertSpan.reduce_size(mFactoryData.dac_cert.len);
+    chip::Crypto::P256PublicKey dacPublicKey;
+
+    ReturnErrorOnFailure(chip::Crypto::ExtractPubkeyFromX509Cert(dacCertSpan, dacPublicKey));
+
+    // In a non-exemplary implementation, the public key is not needed here. It is used here merely because
+    // Crypto::P256Keypair is only (currently) constructable from raw keys if both private/public keys are present.
+    // MutableByteSpan privKeySpan{const_cast<uint8_t*>(reinterpret_cast< const uint8_t *>(mFactoryData.dac_priv_key.data)),
+    // mFactoryData.dac_priv_key.len}; MutableByteSpan pubKeySpan{const_cast<uint8_t*>(reinterpret_cast< const uint8_t
+    // *>(dacPublicKey.Bytes())), dacPublicKey.Length()};
+
+    uint8_t privKeyBuf[600] = { 0 };
+    MutableByteSpan privKeySpan(privKeyBuf);
+    memcpy(privKeySpan.data(), mFactoryData.dac_priv_key.data, mFactoryData.dac_priv_key.len);
+    privKeySpan.reduce_size(mFactoryData.dac_priv_key.len);
+
+    uint8_t pubKeyBuf[600] = { 0 };
+    MutableByteSpan pubKeySpan(pubKeyBuf);
+    memcpy(pubKeySpan.data(), dacPublicKey.Bytes(), dacPublicKey.Length());
+    pubKeySpan.reduce_size(dacPublicKey.Length());
+
+    ReturnErrorOnFailure(LoadKeypairFromRaw(privKeySpan, pubKeySpan, keypair));
+    ReturnErrorOnFailure(keypair.ECDSA_sign_msg(messageToSign.data(), messageToSign.size(), signature));
+
+    return CopySpanToMutableSpan(ByteSpan{ signature.ConstBytes(), signature.Length() }, outSignBuffer);
+}
+CHIP_ERROR FactoryDataProvider::GetSetupDiscriminator(uint16_t & setupDiscriminator)
+{
+    ReturnErrorCodeIf(!mFactoryData.discriminator.data, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND);
+    memcpy(&setupDiscriminator, mFactoryData.discriminator.data, mFactoryData.discriminator.len);
```

Comment:
Needs length check.

Reply:
Somehow missed some of these next few comments when I was first making fixes, added this.